partition_processor.go

```diff
@@ -589,3 +645,43 @@ func (pp *PartitionProcessor) processMessage(ctx context.Context, wg *sync.WaitG
 	msgContext.finish(nil)
 	return nil
 }
+
+// VisitValues iterates over all values in the table and calls the "visit"-callback for the passed name.
+// Optional parameter value can be set, which will just be forwarded to the visitor-function
+func (pp *PartitionProcessor) VisitValues(ctx context.Context, name string, wg *sync.WaitGroup, meta interface{}) error {
+	if pp.table == nil {
+		return fmt.Errorf("cannot visit values in stateless processor")
+	}
+	if _, ok := pp.visitCallbacks[name]; !ok {
+		return fmt.Errorf("unconfigured visit callback. Did you initialize the processor with DefineGroup(..., Visit(%s, ...), ...)?", name)
+	}
+
+	it, err := pp.table.Iterator()
+	if err != nil {
+		return fmt.Errorf("error creating storage iterator")
+	}
+	defer it.Release()
+	for it.Next() {
+
+		// add one that we're going to put into the queue
+		// wg.Done will be called by the visit handler
+		wg.Add(1)
+		select {
+		case <-ctx.Done():
+			// context is closed, so we know that the current visit never entered the channel
+			// so we have to remove it from the waitgroup to avoid the lock.
+			wg.Done()
+			return nil
+		// enqueue the visit
+		case pp.visitInput <- &visit{
+			key:  string(it.Key()),
+			name: name,
+			meta: meta,
+			done: wg.Done,
```

Comment:
@ubunatic
the thing is: it cannot be called concurrently, that's the point of the `select`. If we get a `ctx.Done()`, then the message won't be put into `pp.visitInput`, so it won't be processed by the visitor, so it will never call `wg.Done()` there. In the other case that it ends up in the channel, then it won't go  for the `ctx.Done()` case anymore. 
So actually it should be fine.